### ISSUE TITLE
Feature to only checkpoint and restore memory pages

### ISSUE THREAD

**Issue by DivyanshuSaxena** [checkpoint-restore/criu#1475]
**Is it possible to dump only the memory pages (and the mmap-ed files), while keeping all the other state as it is?**

***Context:***
I am working on a project where I need memory checkpoints for containers. I believe memory footprints are critical in several scenarios and hence, being able to checkpoint memory pages efficiently shall be a very useful feature. From what I understand after having spent a couple of weeks using CRIU with LXC and Docker and after reading the Wiki, checkpointing a container leads to memory dumps as well as other container states (for example, signals, network connections, TTYs, file system mountpoints, etc). While this does solve the goal of getting memory dumps, the checkpointing of all the other container state adds an additional overhead, which I would like to avoid.

I am not very concerned about the integration of this with LXC or Docker or Podman, but I wanted to understand how easy/difficult such a feature would be. I understand that many of these states are interrelated : for example, to restore pages of private file mappings, the file descriptors are needed, and to start the root process in a namespace, the tmpfs mountpoint is (probably) needed.

I thought the Mailing lists were used actively for Development, and hence I raised this as an issue here, under the `question` label. Please let me know if I should move this discussion elsewhere.

**Comment by adrianreber** [comment]
> **Is it possible to dump only the memory pages (and the mmap-ed files), while keeping all the other state as it is?**
> 
> _**Context:**_
> I am working on a project where I need memory checkpoints for containers. I believe memory footprints are critical in several scenarios and hence, being able to checkpoint memory pages efficiently shall be a very useful feature. From what I understand after having spent a couple of weeks using CRIU with LXC and Docker and after reading the Wiki, checkpointing a container leads to memory dumps as well as other container states (for example, signals, network connections, TTYs, file system mountpoints, etc). While this does solve the goal of getting memory dumps, the checkpointing of all the other container state adds an additional overhead, which I would like to avoid.

The overhead of anything besides memory pages is minimal. I would just do a dump and specify `keep-running`. That would be the easiest solution.

> I am not very concerned about the integration of this with LXC or Docker or Podman, but I wanted to understand how easy/difficult such a feature would be. I understand that many of these states are interrelated : for example, to restore pages of private file mappings, the file descriptors are needed, and to start the root process in a namespace, the tmpfs mountpoint is (probably) needed.

As mentioned above. Just dump everything and ignore what you do not need.

**Comment by DivyanshuSaxena** [comment]
Hi @adrianreber, thanks for the quick response. This is very encouraging.
I am utilizing CRIU to speed up slow starts in containers. However, due to the additional state being checkpointed, this is not as efficient as I want it to be. Specifically, I want to make the restores fast.

> The overhead of anything besides memory pages is minimal. I would just do a dump and specify `keep-running`. That would be the easiest solution.

I did a perf events analysis for restore using [FlameGraphs](https://github.com/brendangregg/FlameGraph) and trimming the output to just the CRIU calls looks like this:
[SVG File, viewable and clickable on browser](https://drive.google.com/file/d/1R0lE5XVli_Z5ka_UunVgnyg9ogtDFYRZ/view?usp=sharing)

Specifically, the functions `populate_mnt_ns`, `prepare_cgroups` and `network_<lock/unlock>_internal` seem to be taking significant CPU time. Do you think it will work, if I can remove the cgroup, net and mnt dumps altogether?

**Comment by adrianreber** [comment]
Are you running CRIU in the container? If you are it might be possible to remove the problematic functions. But I think you would be on your own. Not sure anyone has done that before. I am pretty sure the network locking can be removed without any problems (if you are not using `--tcp-established`). Especially as it calls an external tool it probably requires a lot of time.

One of our proposed GSoC projects it replacing the external network locking (via iptables) with a BPF based locking. So not calling an external binary.

I am pretty sure no one has experience with what you are trying to do.

**Comment by DivyanshuSaxena** [comment]
> Are you running CRIU in the container? 
I'm not sure I understand what is meant by running CRIU in the container. I've installed CRIU on the host machine and I am using the docker checkpoint commands. I believe Docker uses the CRIU SWRK mode - is that started from within the container?

I will try to remove the network locking and check if that gives performance improvements.

**Comment by adrianreber** [comment]
> > Are you running CRIU in the container?
> 
> I'm not sure I understand what is meant by running CRIU in the container. I've installed CRIU on the host machine and I am using the docker checkpoint commands. I believe Docker uses the CRIU SWRK mode - is that started from within the container?

No. CRIU is running outside. I was just curious if you are running CRIU manually in the container. Running it in the container would mean that CRIU does not have to deal with namespaces, because you are running it in the namespaces. In your case you are running it outside of the container which means you probably cannot remove most CRIU functionality because it is needed for checkpoint and restoring a container.

> I will try to remove the network locking and check if that gives performance improvements.

Network locking should be removable. I am still sceptical, however, that you will gain much by using a modified version of CRIU. Let us know if you see any improvements.

As you are using CRIU through runc you could try to use a configuration file to reduce the log level from 4 to 0. Formatting log messages is also known to increase CRIU's runtime.
